tests/UKHO.S100PermitService.Common.UnitTests/Encryption/S100CryptTests.cs

```diff
@@ -101,5 +111,92 @@ private List<ProductKeyServiceResponse> GetInvalidProductKeyServiceResponse()
                 }
             ];
         }
+
+        [Test]
+        public void WhenValidMKeyAndUpnInfo_ThenListOfDecryptedHardwareIdIsReturned()
+        {
+            const string FakeDecryptedHardwareId = "86C520323CEA3056B5ED7000F98814CB";
+
+            const string FakeMKey = "validMKey12345678901234567890123";
+
+            A.CallTo(() => _fakeManufacturerKeyService.GetManufacturerKeys(A<string>.Ignored)).Returns(FakeMKey);
+
+            A.CallTo(() => _fakeAesEncryption.Decrypt(A<string>.Ignored, A<string>.Ignored)).Returns(FakeDecryptedHardwareId);
+
+            var result = _s100Crypt.GetDecryptedHardwareIdFromUserPermit(GeUserPermitServiceResponse());
+
+            result.Equals(GetUpnInfo());
+
+            A.CallTo(_fakeLogger).Where(call =>
+                call.Method.Name == "Log"
+                && call.GetArgument<LogLevel>(0) == LogLevel.Information
+                && call.GetArgument<EventId>(1) == EventIds.GetHwIdFromUserPermitStarted.ToEventId()
+                && call.GetArgument<IEnumerable<KeyValuePair<string, object>>>(2)!.ToDictionary(c => c.Key, c => c.Value)["{OriginalFormat}"].ToString() == "Get decrypted hardware id from user permits started"
+            ).MustHaveHappenedOnceExactly();
+
+            A.CallTo(_fakeLogger).Where(call =>
+
+                call.Method.Name == "Log"
+                    && call.GetArgument<LogLevel>(0) == LogLevel.Information
+                    && call.GetArgument<EventId>(1) == EventIds.GetHwIdFromUserPermitCompleted.ToEventId()
+                    && call.GetArgument<IEnumerable<KeyValuePair<string, object>>>(2)!.ToDictionary(c => c.Key, c => c.Value)["{OriginalFormat}"].ToString() == "Get decrypted hardware id from user permits completed"
+                ).MustHaveHappenedOnceExactly();
+        }
+
+        private static List<UpnInfo> GetUpnInfo()
+        {
+            return
+            [
+                new UpnInfo()
+                {
+                    EncryptedHardwareId = "FE5A853DEF9E83C9FFEF5AA001478103",
+                    Upn = "FE5A853DEF9E83C9FFEF5AA001478103DB74C038A1B2C3",
+                    MId = "A1B2C3",
+                    Crc32 = "DB74C038"
+                },
+                new UpnInfo()
+                {
+                    EncryptedHardwareId = "869D4E0E902FA2E1B934A3685E5D0E85",
+                    Upn = "869D4E0E902FA2E1B934A3685E5D0E85C1FDEC8BD4E5F6",
+                    MId = "D4E5F6",
+                    Crc32 = "C1FDEC8B"
+                }
+            ];
+        }
+
+        private static List<UpnInfo> GetUpnInfoWithDecryptedHardwareId()
```

Comment:
Kindly remove this private method if never used in these test cases.

Reply:
Done